sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala

```diff
@@ -274,7 +275,21 @@ object InMemoryFileIndex extends Logging {
     // [SPARK-17599] Prevent InMemoryFileIndex from failing if path doesn't exist
     // Note that statuses only include FileStatus for the files and dirs directly under path,
     // and does not include anything else recursively.
-    val statuses = try fs.listStatus(path) catch {
+    val statuses: Array[FileStatus] = try {
+      fs match {
+        // DistributedFileSystem overrides listLocatedStatus to make 1 single call to namenode
+        // to retrieve the file status with the file block location. The reason to still fallback
+        // to listStatus is because the default implementation would potentially throw a
+        // FileNotFoundException which is better handled by doing the lookups manually below.
+        case _: DistributedFileSystem =>
+          val remoteIter = fs.listLocatedStatus(path)
+          new Iterator[LocatedFileStatus]() {
```

Comment:
Why do you need to wrap the iterator? Should be able to use JavaConversions at worst to make it .asScala.toArray

Reply:
It's not a real iterator unfortunately its a RemoteIterator which which doesn't actually extend the Iterator interface.

Reply:
This seemed cleaner to me than a while loop and an array buffer which is another way to do it. I could do it that way if its preferred or I'm also open to other suggestions

Reply:
I see, makes sense then.